### ISSUE TITLE
Parse TSTP/Prover9 proofs

### ISSUE THREAD

**Issue by newca12** [newca12/gapt#102]
_From [fra...@gmail.com](https://code.google.com/u/108596877348066494139/) on January 10, 2011 13:04:03_

We want to parse Prover9 proofs into gapt. This should be based on the textual proofs that Prover9 outputs, after they are translated using prooftrans (to translate complicated rules into resolution and paramodulation).

This task was started by Daniel. Extending his code, Tomer should write an engine for the ATP prover that instructs ATP to find the proof that prover9 has found.

_Original issue: http://code.google.com/p/gapt/issues/detail?id=102_

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on January 10, 2011 04:27:32_

ATP should have a sub folder in commands for prover9 which will have the following 3 commands:
1) SetProver9ClausesCommand(clauses) replaces setClausesCommand for arbitrary clauses
2) Prover9GetCommand replaces the refinement GetCommands
3) ConfirmAndInsertResolventCommand(resolvent) replaces InsertResolventCommand

The idea is that we will not use an infinite stream for ATP but an exact stream holding the three commands as follows:
We first set a list of clauses taken fro the output of prover9 with place holders for future resolvents.
Then, we use the get command to get the next two clauses from the prover9 list.
In the end we confirm that one of the resolvents returned is equal to the resolvent stated by proof tool (maybe up to renaming or matching) and we replace the right placeholder in the clauses list with the new chosen resolvent.

In the end, if prover9 proof was correct, we would reach a valid proof in our system.

**Owner:** shaolintl  
**Cc:** -shaolintl fraini

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on January 11, 2011 06:34:41_

We would test the programmer reviewer scheme on a task.

**Labels:** Reviewer-Daniel

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 18, 2011 11:23:59_

**Cc:** -fra...@gmail.com  
**Labels:** -Priority-Low -Reviewer-Daniel Priority-Critical

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 19, 2011 02:11:28_

I would suggest parsing TSTP proofs instead of Prover9 proofs. This would allow us to use any current theorem prover...

**Summary:** Parse TSTP proofs  
**Labels:** -Type-Enhancement Type-Task Milestone-Release2.0

**Comment by newca12** [comment]
_From [fra...@gmail.com](https://code.google.com/u/108596877348066494139/) on August 19, 2011 02:24:30_

Is it possible to feasibly transform an arbitrary TSTP proof into a resolution proof? Isn't it possible to use arbitrary rules with TSTP?

If we can't get a resolution proof out of it, parsing such a proof does not help with the problem of implementing CERES, or am I missing something?

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 19, 2011 02:25:18_

is there a tool converting prover9 output to TSTP?

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 19, 2011 02:59:09_

TSTP allows arbitrary (sound) inference rules. To transform a TSTP proof into a resolution proof, we can use Tomer's ATP as follows:

Whenever we have an inference like this in the TSTP proof:

P1  P2  P3 ...  Pn
---------------- arbitrary inference
           C

We call Tomer's ATP on the clauses P1,...,Pn and not C...  Hopefully, ATP is powerful enough to refute this clause set. Tomer's ATP will then output a resolution proof that corresponds to this arbitrary inference...

If we do this for all arbitrary inferences, we will have have many corresponding resolution proofs that can be "easily" glued together.

This technique is called "proof replaying". Its most famous use is probably the sledgehammer approach in Isabelle, which allows Isabelle to call external theorem provers and then replay their TSTP proofs using its internal (simple and not so efficient) theorem prover called Metis.

**Comment by newca12** [comment]
_From [fra...@gmail.com](https://code.google.com/u/108596877348066494139/) on August 19, 2011 03:09:15_

I think the central problem is this:

"Hopefully, ATP is powerful enough to refute this clause set."

which seems doubtful considering our experience with Otterfier: Otter failed to replay simple proofs (eprover refutation of tape clause set), and otter is geared for fully automatic use, while ATP isn't (I believe).

So I would suggest, before replacing the more direct parsing of Prover9 proofs by this, to at least test manually whether ATP can replay (some of) the resolution proofs that have come up in our experiments so far.

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 19, 2011 03:23:40_

Agreed...

Otterfier probably failed to replay eprover's proof because the proof was too coarse-grained. We could use some prover that outputs more fine-grained TSTP proofs, if it exists....

Regarding Tomer's question: is there a tool converting prover9 output to TSTP?

I would suggest contacting Geoff to find this out and also to find out which provers output fine-grained TSTP proofs. The following link is also an interesting place to have a look on proofs output by different provers: http://www.cs.miami.edu/~tptp/cgi-bin/TPTP2T

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 19, 2011 03:38:07_

**Summary:** Parse TSTP/Prover9 proofs

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 19, 2011 07:42:26_

Looking at solutions in the TSTP database, it seems that prover9 does not support the TSTP format but that some other renown provers do, like Vampire, Darwin and E. Most of the provers support TPTP as input so we can do two things:

1) interface Prover9 - from Manos experience it was not so easy but our original idea of replaying prover9 macro steps might make it much easier.
2) implement TSTP - this indeed will support a wide range of provers and as far as I have seen, at least Vampire output first order resolution proofs and does not use other inferences so replaying the proof will be trivial and well in the scope of our ATP. The problem with that is that the format is more extensive (extending TPTP).

What do you think?

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 19, 2011 11:28:10_

I think this is a good idea.

Vampire is also more efficient than Prover9. It is the most efficient prover in most CASC categories, in fact.

Are you sure that Vampire's TSTP proofs are fine-grained resolution proofs?

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 19, 2011 12:02:33_

- Are you sure that Vampire's TSTP proofs are fine-grained resolution proofs?

The two proofs I checked before had only the basic resolution inferences but I checked some more and it seems it is using splitting, definitions, negative normal forms and some similar things which will require a more personal treatment to Vampire than a plain attempt in replaying I think. For example the nnf can only be inferred using the introduction of some extra axioms, not mentioned in the inference of Vampire.

The crucial point is that also our parser of Prover9 output will be based on personally tailored replaying so it seems the work will require anyway two tasks.

1) a parser for Prover9 format or TSTP
2) special commands in ATP which specifically deal with clauses according to extra information. I.e. getting the  nnf inference should be able to create a proof using more than just the upper clause for Vampire and similarily for Prover9.

In short, I think we should implement the TSTP parser and keep looking for the finest grained theorem prover out there so the specialized commands would be as simple as possible.

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 19, 2011 12:05:04_

I forgot to mention Vampire is not open source and documentation consists mainly of publications

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 19, 2011 12:07:33_

**Labels:** Type-Feature

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 19, 2011 12:09:48_

Here is a typical output proof by Vampire:

%----ORIGINAL SYSTEM OUTPUT
% Hi Geoff, go and have some cold beer while I am trying to solve this very hard problem!
% % remaining time: 2999 next slice time: 128
% ott+4_7_bs=off:drc=off:fde=none:lcm=predicate:nwc=1.2:nicw=on:ptb=off:ssec=off:sgo=on:spl=backtracking_116 on ALG036+1
% Refutation found. Thanks to Tanya!
% % SZS status Theorem for ALG036+1
% % SZS output start Proof for ALG036+1
% fof(f435,plain,(
%   $false),
%   inference(subsumption_resolution,[],[f434,f417])).
% fof(f417,plain,(
%   ~sP15(e0)),
%   inference(backward_demodulation,[],[f410,f164])).
% fof(f164,plain,(
%   ~sP15(op(e3,e3))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f410,plain,(
%   e0 = op(e3,e3)),
%   inference(resolution,[],[f28,f362])).
% fof(f362,plain,(
%   sP0),
%   inference(subsumption_resolution,[],[f361,f276])).
% fof(f276,plain,(
%   ~sP1),
%   inference(subsumption_resolution,[],[f273,f38])).
% fof(f38,plain,(
%   e0 != e1),
%   inference(cnf_transformation,[],[f6])).
% fof(f6,axiom,(
%   e0 != e1 & e0 != e2 & e0 != e3 & e1 != e2 & e1 != e3 & e2 != e3),
%   file('/home/graph/tptp/TPTP/Problems/ALG/ALG036+1.p',ax6)).
% fof(f273,plain,(
%   e0 = e1 | ~sP1),
%   inference(backward_demodulation,[],[f271,f21])).
% fof(f21,plain,(
%   op(e0,e0) = e1 | ~sP1),
%   inference(cnf_transformation,[],[f15])).
% fof(f15,plain,(
%   ~sP1 | (op(e0,e0) = e1 & op(e1,e1) = e1 & op(e2,e2) = e1 & op(e3,e3) = e1)),
%   inference(nnf_transformation,[],[f11])).
% fof(f11,plain,(
%   ~sP1 | (op(e0,e0) = e1 & op(e1,e1) = e1 & op(e2,e2) = e1 & op(e3,e3) = e1)),
%   introduced(predicate_definition_introduction,[])).
% fof(f271,plain,(
%   e0 = op(e0,e0)),
%   inference(forward_demodulation,[],[f45,f37])).
% fof(f37,plain,(
%   e0 = unit),
%   inference(cnf_transformation,[],[f4])).
% fof(f4,axiom,(
%   e0 = unit),
%   file('/home/graph/tptp/TPTP/Problems/ALG/ALG036+1.p',ax4)).
% fof(f45,plain,(
%   e0 = op(e0,unit)),
%   inference(cnf_transformation,[],[f2])).
% fof(f2,axiom,(
%   e0 = op(unit,e0) & e0 = op(e0,unit) & op(unit,e1) = e1 & op(e1,unit) = e1 & op(unit,e2) = e2 & op(e2,unit) = e2 & op(unit,e3) = e3 & op(e3,unit) = e3 & (e0 = unit | e1 = unit | e2 = unit | e3 = unit)),
%   file('/home/graph/tptp/TPTP/Problems/ALG/ALG036+1.p',ax2)).
% fof(f361,plain,(
%   sP1 | sP0),
%   inference(resolution,[],[f275,f277])).
% fof(f277,plain,(
%   ~sP2),
%   inference(subsumption_resolution,[],[f274,f39])).
% fof(f39,plain,(
%   e0 != e2),
%   inference(cnf_transformation,[],[f6])).
% fof(f274,plain,(
%   e0 = e2 | ~sP2),
%   inference(backward_demodulation,[],[f271,f17])).
% fof(f17,plain,(
%   op(e0,e0) = e2 | ~sP2),
%   inference(cnf_transformation,[],[f14])).
% fof(f14,plain,(
%   ~sP2 | (op(e0,e0) = e2 & op(e1,e1) = e2 & op(e2,e2) = e2 & op(e3,e3) = e2)),
%   inference(nnf_transformation,[],[f12])).
% fof(f12,plain,(
%   ~sP2 | (op(e0,e0) = e2 & op(e1,e1) = e2 & op(e2,e2) = e2 & op(e3,e3) = e2)),
%   introduced(predicate_definition_introduction,[])).
% fof(f275,plain,(
%   sP2 | sP1 | sP0),
%   inference(subsumption_resolution,[],[f272,f40])).
% fof(f40,plain,(
%   e0 != e3),
%   inference(cnf_transformation,[],[f6])).
% fof(f272,plain,(
%   e0 = e3 | sP2 | sP1 | sP0),
%   inference(backward_demodulation,[],[f271,f33])).
% fof(f33,plain,(
%   op(e0,e0) = e3 | sP2 | sP1 | sP0),
%   inference(cnf_transformation,[],[f13])).
% fof(f13,plain,(
%   (e0 != op(e0,e0) | e0 != op(e1,e1) | e0 != op(e2,e2) | e0 != op(e3,e3)) & (op(e0,e0) != e1 | op(e1,e1) != e1 | op(e2,e2) != e1 | op(e3,e3) != e1) & (op(e0,e0) != e2 | op(e1,e1) != e2 | op(e2,e2) != e2 | op(e3,e3) != e2) & (op(e0,e0) != e3 | op(e1,e1) != e3 | op(e2,e2) != e3 | op(e3,e3) != e3) & (sP0 | sP1 | sP2 | (op(e0,e0) = e3 & op(e1,e1) = e3 & op(e2,e2) = e3 & op(e3,e3) = e3))),
%   inference(definition_folding,[],[f9,f12,f11,f10])).
% fof(f10,plain,(
%   ~sP0 | (e0 = op(e0,e0) & e0 = op(e1,e1) & e0 = op(e2,e2) & e0 = op(e3,e3))),
%   introduced(predicate_definition_introduction,[])).
% fof(f9,plain,(
%   (e0 != op(e0,e0) | e0 != op(e1,e1) | e0 != op(e2,e2) | e0 != op(e3,e3)) & (op(e0,e0) != e1 | op(e1,e1) != e1 | op(e2,e2) != e1 | op(e3,e3) != e1) & (op(e0,e0) != e2 | op(e1,e1) != e2 | op(e2,e2) != e2 | op(e3,e3) != e2) & (op(e0,e0) != e3 | op(e1,e1) != e3 | op(e2,e2) != e3 | op(e3,e3) != e3) & ((e0 = op(e0,e0) & e0 = op(e1,e1) & e0 = op(e2,e2) & e0 = op(e3,e3)) | (op(e0,e0) = e1 & op(e1,e1) = e1 & op(e2,e2) = e1 & op(e3,e3) = e1) | (op(e0,e0) = e2 & op(e1,e1) = e2 & op(e2,e2) = e2 & op(e3,e3) = e2) | (op(e0,e0) = e3 & op(e1,e1) = e3 & op(e2,e2) = e3 & op(e3,e3) = e3))),
%   inference(ennf_transformation,[],[f8])).
% fof(f8,negated_conjecture,(
%   ~((e0 = op(e0,e0) & e0 = op(e1,e1) & e0 = op(e2,e2) & e0 = op(e3,e3)) | (op(e0,e0) = e1 & op(e1,e1) = e1 & op(e2,e2) = e1 & op(e3,e3) = e1) | (op(e0,e0) = e2 & op(e1,e1) = e2 & op(e2,e2) = e2 & op(e3,e3) = e2) | (op(e0,e0) = e3 & op(e1,e1) = e3 & op(e2,e2) = e3 & op(e3,e3) = e3) | ~((e0 = op(e0,e0) & e0 = op(e1,e1) & e0 = op(e2,e2) & e0 = op(e3,e3)) | (op(e0,e0) = e1 & op(e1,e1) = e1 & op(e2,e2) = e1 & op(e3,e3) = e1) | (op(e0,e0) = e2 & op(e1,e1) = e2 & op(e2,e2) = e2 & op(e3,e3) = e2) | (op(e0,e0) = e3 & op(e1,e1) = e3 & op(e2,e2) = e3 & op(e3,e3) = e3)))),
%   inference(negated_conjecture,[],[f7])).
% fof(f7,axiom,(
%   (e0 = op(e0,e0) & e0 = op(e1,e1) & e0 = op(e2,e2) & e0 = op(e3,e3)) | (op(e0,e0) = e1 & op(e1,e1) = e1 & op(e2,e2) = e1 & op(e3,e3) = e1) | (op(e0,e0) = e2 & op(e1,e1) = e2 & op(e2,e2) = e2 & op(e3,e3) = e2) | (op(e0,e0) = e3 & op(e1,e1) = e3 & op(e2,e2) = e3 & op(e3,e3) = e3) | ~((e0 = op(e0,e0) & e0 = op(e1,e1) & e0 = op(e2,e2) & e0 = op(e3,e3)) | (op(e0,e0) = e1 & op(e1,e1) = e1 & op(e2,e2) = e1 & op(e3,e3) = e1) | (op(e0,e0) = e2 & op(e1,e1) = e2 & op(e2,e2) = e2 & op(e3,e3) = e2) | (op(e0,e0) = e3 & op(e1,e1) = e3 & op(e2,e2) = e3 & op(e3,e3) = e3))),
%   file('/home/graph/tptp/TPTP/Problems/ALG/ALG036+1.p',co1)).
% fof(f28,plain,(
%   ~sP0 | e0 = op(e3,e3)),
%   inference(cnf_transformation,[],[f16])).
% fof(f16,plain,(
%   ~sP0 | (e0 = op(e0,e0) & e0 = op(e1,e1) & e0 = op(e2,e2) & e0 = op(e3,e3))),
%   inference(nnf_transformation,[],[f10])).
% fof(f434,plain,(
%   sP15(e0)),
%   inference(subsumption_resolution,[],[f433,f373])).
% fof(f373,plain,(
%   ~sP17(e0)),
%   inference(backward_demodulation,[],[f366,f166])).
% fof(f166,plain,(
%   ~sP17(op(e1,e1))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f366,plain,(
%   e0 = op(e1,e1)),
%   inference(resolution,[],[f26,f362])).
% fof(f26,plain,(
%   ~sP0 | e0 = op(e1,e1)),
%   inference(cnf_transformation,[],[f16])).
% fof(f433,plain,(
%   sP17(e0) | sP15(e0)),
%   inference(subsumption_resolution,[],[f432,f395])).
% fof(f395,plain,(
%   ~sP16(e0)),
%   inference(backward_demodulation,[],[f388,f165])).
% fof(f165,plain,...

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 19, 2011 12:09:48_

...(
%   ~sP16(op(e2,e2))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f388,plain,(
%   e0 = op(e2,e2)),
%   inference(resolution,[],[f27,f362])).
% fof(f27,plain,(
%   ~sP0 | e0 = op(e2,e2)),
%   inference(cnf_transformation,[],[f16])).
% fof(f432,plain,(
%   sP16(e0) | sP17(e0) | sP15(e0)),
%   inference(resolution,[],[f168,f357])).
% fof(f357,plain,(
%   ~sP18(e0)),
%   inference(forward_demodulation,[],[f167,f271])).
% fof(f167,plain,(
%   ~sP18(op(e0,e0))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f168,plain,(
%   sP18(e0) | sP16(e0) | sP17(e0) | sP15(e0)),
%   inference(inequality_splitting,[],[f29,f167,f166,f165,f164])).
% fof(f29,plain,(
%   e0 != op(e3,e3) | e0 != op(e2,e2) | e0 != op(e1,e1) | e0 != op(e0,e0)),
%   inference(cnf_transformation,[],[f13])).
% % SZS output end Proof for ALG036+1
% ------------------------------
% Version: Vampire 1.8 ( revision 1280 )
% Termination reason: Refutation
% Active clauses: 176
% Passive clauses: 222
% Generated clauses: 367
% Final active clauses: 137
% Final passive clauses: 18
% Input formulas: 7
% Initial clauses: 196
% Splitted inequalities: 64
% Fw subsumption resolutions: 37
% Fw demodulations: 72
% Bw demodulations: 54
% Forward subsumptions: 35
% Bw demodulations to eq. taut.: 13
% Binary resolution: 8
% Memory used [KB]: 255
% Time elapsed: 0.022 s
% ------------------------------
% % Success in time 0.029 s

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 19, 2011 23:15:31_

**Labels:** Component-Parsing

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 20, 2011 01:07:12_

I could not see a finer grained output from the other provers. In fact, most of them (the ones that actually solve most of the problems) dont have any proof as output.

I probably repeat what the people who chose Prover9 already knew, but it seems that only vampire and prover9 have a nice output and that prover9 has a finer grained one than Vampire (as far as i could see, I did not read Vampire's guide (as it is non-existent).

I will do a u-turn and if there are no objections i think we should go for prover9 as both the parsing and the replaying are easier (it seems).

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 20, 2011 01:36:55_

Objections to Prover9:

1) With the death of Bill McCune, Prover9 will probably not be maintained anymore...


Advantages of Vampire:

1) It is one of the most efficient provers.
2) It is flexible enough to compete in most CASC categories.


Advantages of TPTP/TSTP:

1) If we had a TSTP parser, I think we could easily display TSTP proofs from any theorem prover in ProofTool. This could attract the attention of the theorem proving community to GAPT...

2) If we want GAPT to be useful for the theorem proving community, and to be more than just a tool used only by us, then we will have to support TPTP/TSTP eventually. And then we might do it already now...

3) Supporting TPTP/TSTP would make GAPT different and better than C++ CERes...


Are you sure that ATP wouldn't manage to cope with Vampire's coarser-grained proofs?

If prover9 is open-source, than an alternative solution would be to modify Prover9's source code, to make it output TSTP, and then parse TSTP in GAPT... I have heard that Bill McCune's code is supposed to be easy to understand.

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 20, 2011 11:48:28_

- Are you sure that ATP wouldn't manage to cope with Vampire's coarser-grained proofs?

they are not coarser-grained really but are stronger than robinson calculus. For example, paramodulation as I have implemented it requires the equality to be a positive literal but in vampire they can deduce the following:

from:
-P15(op(e3,e3)). 
-P16(op(e2,e2)). 
-P17(op(e1,e1)). 
-P18(op(e0,e0)). 
-=(e0,op(e3,e3)) | -=(e0, op(e2,e2)) | -=(e0, op(e1,e1)) | -=(e0, op(e0,e0)). 
deduce:
P18(e0) | P16(e0) | P17(e0) | P15(e0).

They do it via specialized splittings (this one is inequality_splitting) which I have not read about yet (but they should be covered in some of their academic papers)

This is one example where their logic is stronger than Robinson's. As they dont have a guide I am not sure what other things they have.

If we know all the enhancements in Vampire we can add cases for them in ATP.

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 20, 2011 12:17:16_

There is something strange about this example. I think I can produce a model of the 5 premises that is not a model of the conclusion... Maybe there is a typo?

To make things simpler, consider the following similar example:
from:
(1)  -P(a). 
(2)  -Q(b). 
(3)  -=(c,a) | -=(c, b).
deduce:
(4)  P(c) | Q(c).

It is easy to construct a model where 1, 2 and 3 hold, but 4 doesn't...

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 20, 2011 13:01:53_

I fully agree and I dont think there is a typo. Below is the extract from the proof generated by Vampire. There is an option I miss something else.. maybe there is an assumption that everything is P and Q unless otherwise specified. Anyway, this example troubled me also after I wrote it and there are more proofs using this inference, see for example the following proof:

% fof(f7,plain,(
%   ~sP0(difference(a,difference(a,difference(b,difference(b,c)))))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f6,plain,(
%   difference(a,difference(a,difference(b,difference(b,c)))) != difference(difference(a,difference(a,b)),difference(difference(a,difference(a,b)),c))),
%   inference(definition_unfolding,[],[f5,f4,f4,f4,f4])).
% fof(f8,plain,(
%   sP0(difference(difference(a,difference(a,b)),difference(difference(a,difference(a,b)),c)))),
%   inference(inequality_splitting,[],[f6,f7]))


from http://www.cs.miami.edu/~tptp/cgi-bin/SeeTPTP?Category=Solutions&Domain=ALG&File=ALG005-1&System=Vampire---1.8.UNS-Ref.s this is the one i quoted before:

% fof(f164,plain,(
%   ~sP15(op(e3,e3))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f165,plain,(
%   ~sP16(op(e2,e2))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f166,plain,(
%   ~sP17(op(e1,e1))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f167,plain,(
%   ~sP18(op(e0,e0))),
%   introduced(inequality_splitting_name_introduction,[])).
% fof(f29,plain,(% fof(f165,plain,(
%   ~sP16(op(e2,e2))),
%   introduced(inequality_splitting_name_introduction,[])).
%   e0 != op(e3,e3) | e0 != op(e2,e2) | e0 != op(e1,e1) | e0 != op(e0,e0)),
%   inference(cnf_transformation,[],[f13])).
% fof(f168,plain,(
%   sP18(e0) | sP16(e0) | sP17(e0) | sP15(e0)),
%   inference(inequality_splitting,[],[f29,f167,f166,f165,f164])).

There must be a very simple answer

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 20, 2011 13:08:12_

Maybe what I wrote above as a joke:
maybe there is an assumption that everything is P and Q unless otherwise specified
is really the case i.e. they compactize the domain by specifying only finitely many negative witnesses. Like ~(0>0) for natural numbers?

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 21, 2011 02:23:29_

I would suggest contacting Andrei Voronkov to find out where this inequality splitting rule is explained...

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 21, 2011 02:27:32_

i will check first the papers they wrote. Maybe it is explained there.

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 21, 2011 03:39:57_

the splitting rules can be turned of using:

--forced_options
propositional_to_bdd=off:splitting=off:equality_proxy=off:
general_splitting=off:inequality_splitting=0:naming=0

I wanted to see the output of the proof with these features turned off but nowhere can one find even the usage of the prover. Does anybody know how to use this prover given a TPTP input file?

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 21, 2011 03:44:37_

I don't know. There was a tutorial about Vampire in the last CADE, but unfortunately I could not attend, because it coincided with one of my talks.

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 21, 2011 03:48:10_

Laura Kovacs, who used to work with Andrei Voronkov in Manchester, is now here at TU-Vienna (in the Informatik department, but not in the institut für Computersprachen, I think)... I think she could help you with this...

**Comment by newca12** [comment]
_From [lambda.x...@gmail.com](https://code.google.com/u/103054511164786171111/) on August 21, 2011 14:03:43_

Laura Kovacs is in the other working group ( http://www.complang.tuwien.ac.at/lkovacs/).. . i wanted to contact her for the integration of theories into vampire anyway. Perhaps we could ask for for a meeting?

**Comment by newca12** [comment]
_From [fra...@gmail.com](https://code.google.com/u/108596877348066494139/) on August 23, 2011 06:41:17_

Issue 105 has been merged into this issue.

**Cc:** mrukh...@gmail.com shaoli...@gmail.com

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 24, 2011 04:09:48_

Issue 64 has been merged into this issue.

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 28, 2011 00:27:27_

I had a look on the tools prover9 supplies and among there there are:
- tptp to ladr tool
- expand tool which removes all extra inferences (like hyper resolution, deletions and demodulation)tptp_to_ladr
- export to a simple xml

I think we should therefore support parsing of tptp (cnf for starts) and integrate prover9 within the system:

1) tptp-cnf parser (to implement))
2) apply tptp_to_ladr, prover9, prooftrans xml expand (to integrate)
3) parse xml to our ds (to implement)
4) minimal replay of the proof within our system (to implement)

**Cc:** bruno...@gmail.com fra...@gmail.com

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 28, 2011 00:40:11_

of course, in all cases a tptp parser is not required (at least for the application of a 3rd party prover). my mistake.

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 28, 2011 00:57:57_

Are you suggesting that we should first generate XML, before parsing into our data structures?

Could you also explain step 2 better?

Daniel and I talked a bit about replaying of arbitrary proofs within GAPT last tuesday. Remind me to share our ideas with you next time.

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 28, 2011 01:13:41_

Right now prover9 can take a tptp problem and generate an xml containing very simple inferences only. In step 2 the idea is to obtain this xml. The next 2 steps will convert it into a proof in our system.

**Comment by newca12** [comment]
_From [fra...@gmail.com](https://code.google.com/u/108596877348066494139/) on August 28, 2011 01:29:28_

Note that there is already a Prover9.scala which uses our TPTP export code and tptp_to_ladr to generate Prover9 input (i.e. in LADR format) from a set of clauses in our data structures. Prover9 output is not handled there at all yet.

Also, be sure to test prooftrans on a few examples: if I recall correctly, while it expands many inferences, I don't think that the result is always a proof based purely on resolution and paramodulation.

I have no experience at all with the prover9 XML proof format. Do you have an example of that?

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 28, 2011 01:39:03_

I have seen this file and I think to include it in a ATP command so the user will not handle anything outside of our data structures in the interactive shell.

I looked on one example so far (attached) and it looked good (except some addtional very simple transformations). not a very difficult problem though.

**Attachment:** [alg002-1.xml](http://code.google.com/p/gapt/issues/detail?id=102#hc37)

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on August 28, 2011 01:48:15_

Interestingly, since clauses are flat data structures, XML turns out to be not so bad to represent them...

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on August 30, 2011 07:01:22_

**Status:** Started  
**Blockedon:** 117

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on October 10, 2011 13:10:58_

Hi, Tomer!
I have seen that you made many commits related to this issue recently. Thanks for that! 
I still need to find time to have a closer look at those commits, though.
In the meanwhile, how is the status of this issue right now? If possible, could you summarize it before your holidays?

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on October 10, 2011 13:31:54_

I am parsing right now prover9 output converted into xml and into a simple proof using prooftrans xml expand from the prover9 tools.

Prover9 proved to be the easiest prover to replay the proofs of.

**Comment by newca12** [comment]
_From [bruno...@gmail.com](https://code.google.com/u/105016684496602932564/) on October 11, 2011 00:00:45_

ok! thanks!

**Comment by newca12** [comment]
_From [shaoli...@gmail.com](https://code.google.com/u/113190107447576027220/) on November 10, 2011 06:20:15_

done. tested on the tape proof. there might be prover9 calls that were not taken care.

**Status:** Fixed
